src/vs/editor/common/modes/textToHtmlTokenizer.ts

```diff
@@ -8,11 +8,115 @@ import * as strings from 'vs/base/common/strings';
 import { IState, ITokenizationSupport, TokenizationRegistry, LanguageId } from 'vs/editor/common/modes';
 import { NULL_STATE, nullTokenize2 } from 'vs/editor/common/modes/nullMode';
 import { LineTokens } from 'vs/editor/common/core/lineTokens';
+import { CharacterMapping } from 'vs/editor/common/viewLayout/viewLineRenderer';
+import { CharCode } from 'vs/base/common/charCode';
+import { ViewLineToken } from 'vs/editor/common/core/viewLineToken';
 
 export function tokenizeToString(text: string, languageId: string): string {
 	return _tokenizeToString(text, _getSafeTokenizationSupport(languageId));
 }
 
+export function tokenizeLineToHTML(text: string, viewLineTokens: ViewLineToken[], rules: { [key: string]: string }, options: { startOffset: number, endOffset: number, tabSize: number, containsRTL: boolean }): string {
+	let tabSize = options.tabSize;
+	let containsRTL = options.containsRTL;
+
+	let result = `<div>`;
+	const characterMapping = new CharacterMapping(text.length + 1, viewLineTokens.length);
+
+	let charIndex = options.startOffset;
+	let tabsCharDelta = 0;
+	let charOffsetInPart = 0;
+
+	for (let tokenIndex = 0, lenJ = viewLineTokens.length; tokenIndex < lenJ; tokenIndex++) {
+		const token = viewLineTokens[tokenIndex];
+		const tokenEndIndex = token.endIndex;
+
+		if (token.endIndex < options.startOffset) {
+			continue;
+		}
+
+		const tokenType = token.type;
+		let partContentCnt = 0;
```

Comment:
since there's no need to build up character mapping, there's also no need to keep track of partContentCnt